### PR TITLE
Implemented Swimming

### PR DESCRIPTION
**Pull request by buchwasa**
## Introduction
This pull request aims to add the missing swimming functionality that was added back in 1.4.

### API changes
Added
Living->swimming (protected)
Living->setSwimming(bool $value) : void;
PlayerToggleSwimEvent
Player->toggleSwim(bool $swim) : bool
### Behavioural changes
Hunger now takes account for swimming, 0.015, as stated in the comment.

## Backwards compatibility
This is backwards compatible with current API.

## Follow-up
I'm not too certain how well I like the setSwimming() function.

## Tests
I tested by going underwater and sprinting and going to the surface for air. I've also included a video to show it in action.
[Streamable Video](https://streamable.com/a97q6n)

### CODE REVIEW

File: src/entity/Living.php [hunk 1]
```diff
@@ -221,6 +223,22 @@ public function setSprinting(bool $value = true) : void{
 		}
 	}
 
+	public function setSwimming(bool $value = true) : void{
+		if($value !== $this->isSwimming()) {
+			$this->swimming = $value;
+			if($this->swimming){
+				$this->height = $this->eyeHeight = $this->width;
```
Comment (dktapps) [inline src/entity/Living.php:230]: This will mess up the bounding box irreparably.

File: src/entity/Living.php [hunk 2]
```diff
@@ -221,6 +223,22 @@ public function setSprinting(bool $value = true) : void{
...
+				$this->recalculateBoundingBox();
+			}else{
+				$this->setScale($this->getScale());
```
Comment (dktapps) [inline src/entity/Living.php:233]: Not sure what this is supposed to do.

File: src/event/player/PlayerToggleSwimEvent.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+	/** @var bool */
+	protected $isSwimming;
+
+	public function __construct(Player $player, bool $isSwimming){
+		$this->player = $player;
+		$this->isSwimming = $isSwimming;
+	}
+
+	public function isFlying() : bool{
```
Comment (GiantQuartz) [inline src/event/player/PlayerToggleSwimEvent.php:41]: Is this name intentional? It doesn't seem to fit with the context.

**Review by dresnite — CHANGES_REQUESTED** [review]

File: src/event/player/PlayerToggleSwimEvent.php (hunk 3 above)
Comment (buchwasa) [reply src/event/player/PlayerToggleSwimEvent.php:41]: oops thanks for catching my laziness

File: src/entity/Living.php [hunk 4]
```diff
@@ -568,6 +600,14 @@ protected function entityBaseTick(int $tickDiff = 1) : bool{
 			if($this->doAirSupplyTick($tickDiff)){
 				$hasUpdate = true;
 			}
+
+			if(
+				$this->isSwimming() && !$this->isUnderwater() &&
```
Comment (thebigsmileXD) [inline src/entity/Living.php:605]: isUnderwater() should finally be changed to check if the bounding box is submerged.
I suggest a change in syntax as follows:
`function isUnderwater(bool $checkSubmerged = false)`
This way old behaviour will be kept by default, but also allowing checks like the ones below to be removed

File: src/entity/Living.php [hunk 5]
```diff
@@ -237,15 +237,14 @@ public function setSwimming(bool $value = true) : void{
...
 			$this->swimming = $value;
 			if($this->swimming){
-				$this->originalWidth = $this->width;
-				$this->originalHeight = $this->height;
-				$this->originalEyeHeight = $this->eyeHeight;
-				$this->height = $this->eyeHeight = $this->width;
+				$width = $this->size->getWidth();
+				$this->originalWidth = $width;
+				$this->originalHeight = $this->size->getHeight();
+				$this->originalEyeHeight = $this->size->getEyeHeight();
```
Comment (dktapps) [inline src/entity/Living.php:243]: why not just make an originalSize field? ...

Comment (buchwasa) [reply src/entity/Living.php:243]: great idea dktapps

**Comment by dktapps** [comment]
Actually, I think originalSize isn't needed; getInitialSize()->scale() can be used, like this: https://github.com/pmmp/PocketMine-MP/blob/master/src/entity/Entity.php#L297

You might also want to adjust setScale() to account for the changes in this PR.

**Pull request closed without merging**